Add pvl instruction tests

mlsteele opened the pull request:
Adds some tests for each instruction. These tests should cover normal operation of pvl (both succeeding and failing instructions). They do not cover the darker corners of malformed pvl, running dns proofs, nor malformed top-level pvl.

Other goodies:
- add mocking mode to `StubAPIEngine`
- fixed crash on negative indices in selectors
- added `func pyindex` so that negative indices act as specified

r? @maxtaco @oconnor663

maxtaco commented:
@mlsteele see C/I failures, I think they're real failures introduced with this change.

maxtaco commented on go/libkb/api_stub.go:
```diff
@@ -5,31 +5,146 @@
+		(okhtml || (restype != XAPIResHTML)) &&
+		(oktext || (restype != XAPIResText))
+	if !ok {
+		fmt.Printf("endpoint: %v\n", arg.Endpoint)
```
is it possible to use logging for this?

maxtaco commented on go/libkb/api_stub.go:
```diff
@@ -5,31 +5,146 @@
 }
 
 func (e *StubAPIEngine) Get(arg APIArg) (res *ExternalAPIRes, err error) {
+	if e.mocking {
```
in what case would this StubAPI be linked to from `G` but we would want mocking off?

maxtaco commented on go/libkb/api_stub.go:
```diff
@@ -5,31 +5,146 @@
 func (e *StubAPIEngine) GetText(arg APIArg) (*ExternalTextRes, error) {
-	if body, ok := apiStubs[arg.Endpoint]; ok {
-		return &ExternalTextRes{Body: body, HTTPStatus: 200}, nil
+	if e.mocking {
```
the CI failure is that the old mode of behavior broke here.

maxtaco commented on go/libkb/api_stub.go:
```diff
@@ -5,31 +5,146 @@
+	if len(e.calls) > 1 {
+		return fmt.Errorf("stub api called more than once")
+	}
+	call := e.calls[0]
```
do you need to reset after this?

maxtaco commented on go/libkb/api_stub.go:
```diff
@@ -5,31 +5,146 @@
 	return &StubAPIEngine{
 		ExternalAPIEngine: &ExternalAPIEngine{BaseAPIEngine{clients: make(map[int]*Client)}},
+		mocking:           false,
+		expectingJSON:     make(map[string]*ExternalAPIRes),
```
would id make more sense to have one map that maps a string a struct that has 3 fields?

maxtaco commented on go/pvl/interp.go:
```diff
@@ -95,22 +95,42 @@ var Steps = map[CommandName]Step{
 }
 
+// ProofInfo contains all the data about a proof PVL needs to check it.
+// It can be derived from a RemoteProofChainLink and SigHint.
```
Ok, great!

maxtaco commented on go/libkb/api_stub.go:
```diff
@@ -5,31 +5,146 @@
 }
 
 func (e *StubAPIEngine) GetText(arg APIArg) (*ExternalTextRes, error) {
-	if body, ok := apiStubs[arg.Endpoint]; ok {
```
you dropped this!

maxtaco commented on go/pvl/interp_test.go:
```diff
@@ -1,4 +1,647 @@
+		name:      "AssertRegexMatch-content-ok",
+		proofinfo: info1,
+		prepvl: map[keybase1.ProofType]string{
+			keybase1.ProofType_REDDIT: `[
```
should this be a twitter proof, and not a reddit proof?

mlsteele replied:
These are all fake anyway. I tried to vary which ProofType is used just to encourage bugs in any future service-specific code to show up.

mlsteele commented:
- restored libkb/api_stub.go (should fix test failures)
- added pvl/api_stub.go
- got rid of the `mocking` flag (always on for `pvl/api_stub.go`)
- removed accidental printfs
- made stub api expectations a map to a triplet-struct

maxtaco commented:
I'm a 👍, great work @mlsteele !!!!
